src/Emogrifier/HtmlProcessor/HtmlPruner.php

```diff
@@ -41,4 +41,47 @@ public function removeInvisibleNodes()
 
         return $this;
     }
+
+    /**
+     * Removes classes that are no longer required (e.g. because there are no longer any CSS rules that reference them)
+     * from `class` attributes.
+     *
+     * Note that this does not inspect the CSS, but expects to be provided with a list of classes that are still in use.
+     *
+     * This method also has the (presumably beneficial) side-effect of minifying (removing superfluous whitespace from)
+     * `class` attributes.
+     *
+     * @param string[] $classesToKeep list of class names that should not be removed
+     *
+     * @return self fluent interface
+     */
+    public function removeRedundantClasses(array $classesToKeep)
+    {
+        $nodesWithClassAttribute = $this->xPath->query('//*[@class]');
+
+        if ($classesToKeep !== []) {
+            // https://stackoverflow.com/questions/6329211/php-array-intersect-efficiency
+            // It's more efficient to invert the array and use `array_intersect_key()` when doing many intersections.
+            $classesToKeepAsKeys = \array_flip($classesToKeep);
+
+            foreach ($nodesWithClassAttribute as $node) {
+                $nodeClasses = \preg_split('/\\s++/', \trim($node->getAttribute('class')));
```

Comment:
What does the second plus sign mean?

Reply:
It's the [possessive quantifier](https://www.regular-expressions.info/possessive.html) - it tells the regex engine not to worry about backtracking, so makes it more optimal and helps mitigate against [catastrophic failure](https://www.regular-expressions.info/catastrophic.html).

If the regex was instead `/\s+\n/`, the possessive quantifier could not be used, as the regex engine may need to backtrack to see if one of the whitespace characters matched was in fact a newline - however that regex could be rewritten as `/[^\S\n]++\n/` to ensure the first part won't match a newline, then the possessive quantifier can be used.  Other examples include rewriting `/\(.*\)/` as `/\([^\)]*+\)` (matching something in brackets) - with `/\(.*\)/` the regex engine could have to backtrack a long way to find the closing bracket (or lack of).